### ISSUE TITLE
Changing a class can leave an obsolete metaclass behind

### ISSUE THREAD

**Issue by daniels220** [dolphinsmalltalk/Dolphin#258]
There's a comment in ClassBuilderTests>>testReshapeClass, stating "Issue: Dolphin's ClassBuilder fails here because it does not become over the metaclasses, only the classes.". Is there any reason _not_ to just add `oldClass class oneWayBecome: newClass class` right before the `#become:`/`#oneWayBecome:` pair that swaps the classes at the bottom of `#mutateClass:toBeASubclassOf:`? If that seems reasonable, I'll change the tests to remove the comment and `#runKnownBug:test:` calls and make a PR.

(I've encountered problems due to this, when there is a direct reference to a metaclass held in a block stored in a class variable, or a compile-time literal, and I change the class and end up with two metaclasses, one of which is the actual class of the one-and-only instance class, the other of which is an obsolete hanger-on.)

**Comment by blairmcg** [comment]
Quite possibly, needs a little thought...

**Comment by blairmcg** [comment]
The handling of metaclasses at the top of `#mutateClass:toBeASubclassOf:` makes me think more than a little thought is needed. Something else that is wrong is that the metaclass methods are not recompiled.

```
Date subclass: #Wibble
	instanceVariableNames: ''
	classVariableNames: ''
	poolDictionaries: ''
	classInstanceVariableNames: ''!

!Wibble methodsFor!
blah
	^DaysInMonth! !
!Wibble categoriesFor: #blah!public! !

!Wibble class methodsFor!
blah
	^DaysInMonth! !
!Wibble class categoriesFor: #blah!public! !

Wibble blah "Works as expected"!

Wibble new blah "Works as expected"!

Association subclass: #Wibble
	instanceVariableNames: ''
	classVariableNames: ''
	poolDictionaries: ''
	classInstanceVariableNames: ''!

Wibble blah "Works - not expected"!

Wibble new blah "Fails as expected"!
```

**Comment by daniels220** [comment]
Thinking about it some myself, indeed the simple #oneWayBecome: approach does not quite work. It will solve this problem, but introduce a subtle new bug—the identityHash of the metaclass will change, breaking any hashed collections it may be a part of. Unfortunately...well, this just gets weirder and weirder. I tried just swapping it in the same way as the instanceClass, with code like:

```
oldMeta := oldClass class.
newMeta := newClass class.
oldMeta removeFromSuper.
oldClass removeFromSuper.
oldClass become: newClass.
newClass oneWayBecome: oldClass.
oldMeta become: newMeta.
newMeta oneWayBecome: oldMeta.
```

and this broke the boot process with an `Index 14 is out of bounds` error. I narrowed it down to the following scenario involving class instance variables (evaluate after applying the change above):

```
"Make it easy to evaluate this whole mess repeatedly."
Smalltalk at: #ClassA ifPresent: [:cls | cls removeFromSystem].
Object
	subclass: #ClassA
	instanceVariableNames: ''
	classVariableNames: ''
	poolDictionaries: ''
	classInstanceVariableNames: 'cVarA'.
Object
	subclass: #ClassB
	instanceVariableNames: ''
	classVariableNames: ''
	poolDictionaries: ''
	classInstanceVariableNames: ''.
"First half of #subclass:...classInstanceVariableNames:"
(Smalltalk at: #ClassA)
	subclass: #ClassB
	instanceVariableNames: ''
	classVariableNames: ''
	poolDictionaries: ''.
"Everything *appears* fine at this point:"
(Smalltalk at: #ClassB) instVarAt: 14 put: #foo.
self assert: [((Smalltalk at: #ClassB) instVarAt: 14) == #foo].
"Buuuuut no:"
(Smalltalk at: #ClassB) class instanceVariableNames: 'cVarB'.
```

You should get the error in the very last statement, trying to copy the value of `cVarA` from the old `ClassB` to the new. BUT! This is a Heisenbug! (Trying to debug through the last statement makes the problem not appear.) Long story short, simply causing a context switch (via e.g. `Processor sleep: 0`) in between the calls to Class>>subclass:...classVariableNames: and Metaclass>>instanceVariableNames: prevents the problem from manifesting.

It's worth noting, though, that once it _has_, it is persistent—opening a debugger on the error and re-evaluating `self instVarAt: 14` gives the same error. At this point:

* `oldInstance basicSize` returns `SmallInteger maximum`.
* There are _three_ `ClassB`s—`ClassA class primAllSubinstances` answers four objects.

So...this is looking like either a VM bug (is `#oneWayBecome:` not truly atomic somehow?), or something the image should be telling the VM about that it isn't. Any thoughts?

**Comment by blairmcg** [comment]
Yes, you certainly need the become:/oneWayBcome: pair. As you realised, that trick is necessary to preserve the identity of the original object when replacing it with a new one. 

#oneWayBecome: is a synchronous operation, but it sounds to me like may have a bug in it. The class pointer for an object is stored in the object table entry, not in the object body. Perhaps it is not updating those correctly, although you would expect that to be causing a problem with moving object instances over to new classes too.

Might be worth running your test script on a debug VM to see if any assertions fire.

**Comment by daniels220** [comment]
It is frustrating, but not entirely surprising, that the problem simply does not appear when running that script in a debug VM. There is some output to the debug trace device:

```
GC WARNING: 10435B40/4 (size 60) has freed class 104358F0/210319
GC WARNING: 1045EBA0/27 (size 56) has freed class 104354C0/210252
GC: 8 Array's were deallocated
GC: 6 String's were deallocated
GC: 1 Symbol's were deallocated
GC: 1 ByteArray's were deallocated
GC: 4 MethodDictionary's were deallocated
GC: 2 Metaclass's were deallocated
GC: 2 BlockClosure's were deallocated
GC: 1 CompiledExpression's were deallocated
GC: 2 GUID's were deallocated
GC: 1 objects of a free'd class (104354C0(210252) were deallocated
GC: 1 objects of a free'd class (104358F0(210319) were deallocated
GC: Completed, 29 objects reclaimed, 0 queued for finalization, 0 weak lose elements
```

The warnings look a little scary, and definitely related, but on further thought they kinda make sense given that #oneWayBecome: is involved... I can't say I have a firm enough grasp on the gyrations involved to say for sure. (Worth noting that 56 and 60 are the byte sizes of ClassA and ClassB, respectively.)

**Comment by blairmcg** [comment]
Hmmm. Well that's useful - there is a clue there in that those warnings should never appear. This suggests that the oneWayBecome operation is not working correctly as you originally suspected.

**Comment by blairmcg** [comment]
@daniels220, after a substantial amount of time elapsed, I found the reason for the problem you described as:

>and this broke the boot process with an Index 14 is out of bounds error. I narrowed it down to the following scenario involving class instance variables (evaluate after applying the change above):
> ....
> So...this is looking like either a VM bug (is `#oneWayBecome:` not truly atomic somehow?), or something the image should be telling the VM about that it isn't. Any thoughts?

Just by coincidence some other changes I made caused this issue to start showing up as a failure of one of the ClassBuilder tests - not always, and it didn't happen on the CI build, but it did happen when I ran the tests from the command line repeatedly. So I looked into it, and it turned out to be because of old invalid instances created by the ClassBuilder mutation that should have become unreachable garbage (the `oneWayBecome: `after the `become:` swapping old with new sees to that) but were being picked up by the subsequent scan for instances if it happened soon enough. This would actually result in invalid old instances being returned as instances of the new class, and of course they may have the wrong size, so two or more ClassBuilder mutations in quick succession might fail after the first. This didn't happen with the debug VM because it deliberately flushed the internal tracking structure very frequently - I actually found the problem by turning that off.

The fix was in f5cbee3fc6037d6912945c97f7d03a155deeb386.

I haven't attempted to fix the original issue described in your first report. It may be that your solution will work now, but I've not tried it yet.
